builtin/joined_dataset.cc

```diff
@@ -360,6 +239,223 @@ struct JoinedDataset::Itl
+                queryExpression.clauses.push_back(rowExpression);
+
+                auto generator = dataset.queryBasic
+                (context, queryExpression, side.when, *condition, side.orderBy,
+                 0, -1, true /* allowParallel */);
+                auto rows = generator(-1);
+            
+                if (debug)
+                    cerr << "got rows " << jsonEncode(rows) << endl;
+                
+                // Now we extract all values 
+                std::vector<std::tuple<ExpressionValue, RowName, RowHash> > sorted;
+
+                for (auto & r: rows) {
+                    ExcAssertEqual(r.columns.size(), 1);
+
+                    const ExpressionValue & embedding = std::get<1>(r.columns[0]);
+                    if (outer)
+                    {
+                        const ExpressionValue & condition = embedding.getField(1);
+                        if (!condition.asBool())
+                        {
+                            recordOuterRow(r.rowName, r.rowHash);
+                            continue;
+                        }
+                    }
+
+                    const ExpressionValue & value = embedding.getField(0);
+                    sorted.emplace_back(value, r.rowName, r.rowHash);
+                }
+
+                std::sort(sorted.begin(), sorted.end());
+
+                return sorted;
+            };
+
+        auto recordOuterLeft = [&] ( const RowName& rowName, const RowHash& rowHash )
+        {
+            recordJoinRow(rowName, rowHash, RowName(), RowHash());
+        };
+
+        auto recordOuterRight = [&] ( const RowName& rowName, const RowHash& rowHash  )
+        {
+            recordJoinRow( RowName(), RowHash(), rowName, rowHash);
+        };
+
+        std::vector<std::tuple<ExpressionValue, RowName, RowHash> > leftRows, rightRows;
+
+        leftRows = runSide(condition.left, *left.dataset, outerLeft, recordOuterLeft);
+        rightRows = runSide(condition.right, *right.dataset, outerRight, recordOuterRight);
+
+        switch (condition.style) {
+        case AnnotatedJoinCondition::CROSS_JOIN: {
+            // Join with no restrictions on the joined column
+            if (leftRows.size() * rightRows.size() > 100000000) {
+                throw HttpReturnException(400, "Cross join too big: cowardly refusing to materialize row IDs for a dataset with > 100,000,000 rows",
+                                          "leftSize", leftRows.size(),
+                                          "rightSize", rightRows.size());//,
+                                          //"joinOn", joinConfig.on);
+            }
+            break;
+        }
+        case AnnotatedJoinCondition::EQUIJOIN: {
+            // Join on f(leftrow) = f(rightrow)
+            break;
+        }
+        default:
+            throw HttpReturnException(400, "Join expression requires an equality operator; needs to be in the form f(left) = f(right)",
```

Comment:
If I am not mistaken the equality expression is required only for inner and outer joins.  We should modify the error message to be more specific.